sources/PTYSession.m

```diff
@@ -13343,23 +13343,84 @@ - (void)screenDidUpdateReturnCodeForMark:(id<VT100ScreenMarkReading>)mark
     [self screenNeedsRedraw];
 }
 
-- (void)screenPostUserNotification:(NSString * _Nonnull)message {
+- (void)screenPostUserNotification:(NSString * _Nonnull)message rich:(BOOL)rich {
     if (![self shouldPostTerminalGeneratedAlert]) {
         DLog(@"Declining to allow terminal to post user notification %@", message);
         return;
     }
     DLog(@"Terminal posting user notification %@", message);
     [self incrementBadge];
-    NSString *description = [NSString stringWithFormat:@"Session %@ #%d: %@",
-                             [[self name] removingHTMLFromTabTitleIfNeeded],
-                             [_delegate tabNumber],
-                             message];
-    [[iTermNotificationController sharedInstance]
-     notify:@"Alert"
-     withDescription:description
-     windowIndex:[self screenWindowIndex]
-     tabIndex:[self screenTabIndex]
-     viewIndex:[self screenViewIndex]];
+
+    iTermNotificationController* controller = [iTermNotificationController sharedInstance];
+
+    if (rich) {
+        NSArray *parts = [message componentsSeparatedByString:@"|"];
+
+        if ([parts count] != 5) {
+            DLog(@"Incorrect number of parts in rich notification, %lu vs 5: %@",
+                    [parts count], message);
+            return;
+        }
+
+        NSString *title = nil;
+        NSString *subtitle = nil;
+        NSString *description = nil;
+        NSString *image = nil;
+        NSString *sound = nil;
+
+        NSMutableArray *decoded_parts = [[NSMutableArray alloc] init];
```

Comment:
Avoid using _ in variable names, preferring camelCase. Also this file is manual retain-release so this is a memory leak. This would be correct:

```
NSMutableArray<NSString *> *decodedParts = [NSMutableArray];
```

Reply:
Thanks for pointing out the memory leak here!  In any event, this code no longer exists in my revised version.